aws-datastore/src/main/java/com/amplifyframework/datastore/DataStoreConfiguration.java

```diff
@@ -0,0 +1,349 @@
+                    case SYNC_MAX_RECORDS:
+                        builder.syncMaxRecords(pluginJson.getInt(ConfigKey.SYNC_MAX_RECORDS.toString()));
+                        break;
+                    case SYNC_PAGE_SIZE:
+                        builder.syncPageSize(pluginJson.getInt(ConfigKey.SYNC_PAGE_SIZE.toString()));
+                        break;
+                    default:
+                        throw new IllegalArgumentException("Unsupported config key = " + configKey.toString());
+                }
+            } catch (JSONException jsonException) {
+                throw new DataStoreException(
+                        "Issue encountered while parsing configuration JSON",
+                        jsonException, "Ensure your amplifyconfiguration.json is valid."
+                );
+            }
+        }
+
+        return builder;
+    }
+
+    /**
+     * Creates an {@link DataStoreConfiguration} which uses all default values.
+     * @return A default {@link DataStoreConfiguration}
+     */
+    @NonNull
+    public static DataStoreConfiguration defaults() {
+        return builder()
+            .syncIntervalInMinutes(DEFAULT_SYNC_INTERVAL_MINUTES)
+            .syncPageSize(DEFAULT_SYNC_PAGE_SIZE)
+            .syncMaxRecords(DEFAULT_SYNC_MAX_RECORDS)
+            .build();
+    }
+
+    /**
+     * Gets the data store error handler.
+     * @return Data store error handler.
+     */
+    @NonNull
+    public DataStoreErrorHandler getDataStoreErrorHandler() {
+        return this.dataStoreErrorHandler;
+    }
+
+    /**
+     * Gets the data store conflict handler.
+     * @return Data store conflict handler
+     */
+    @NonNull
+    public DataStoreConflictHandler getDataStoreConflictHandler() {
+        return this.dataStoreConflictHandler;
+    }
+
+    /**
+     * Get the sync interval. The sync interval is the amount of time after a base sync, during which
+     * the optimized delta-sync may be requested, instead of a full base sync.
+     * @return The sync interval
+     */
+    @IntRange(from = 0)
+    public Long getSyncIntervalMs() {
+        return this.syncIntervalMs == null ? DEFAULT_SYNC_INTERVAL_MS : this.syncIntervalMs;
+    }
+
+    /**
+     * Gets the maximum number of records that the client wants to process, while it is requesting
+     * a base/delta sync operation from AppSync.
+     * @return The max number of records to process from AppSync.
+     */
+    @IntRange(from = 0)
+    public Integer getSyncMaxRecords() {
+        return this.syncMaxRecords == null ? DEFAULT_SYNC_MAX_RECORDS : this.syncMaxRecords;
+    }
+
+    /**
+     * Gets the number of items that should be requested in page, from AppSync, during
+     * a sync operation.
+     * @return Desired size of a page of results from an AppSync sync response
+     */
+    @IntRange(from = 0)
+    public Integer getSyncPageSize() {
+        return this.syncPageSize == null ? DEFAULT_SYNC_PAGE_SIZE : this.syncPageSize;
```

Comment:
Can you put this construction stuff in the `build()` so that the `DataStoreConfiguration` doesn't have any logic, it just returned a single, final, stored instance value?

(Same for others with this pattern.)

Reply:
Changes made in the [latest commit](https://github.com/aws-amplify/amplify-android/pull/401/commits/6c3c4c5522dcd21c055d511c03afb970e64bede5)

With the latest commit, the following rules should apply:
* Read settings from config file (if any)
* Read settings from user-provided POJO (if any). Apply any non-null values
* Fill in any missing fields with default values if instructed to do so (More on that below)

## Changes to `Builder`
* Moved most of the logic out of the `DataStoreConfiguration` methods
* Added a constructor that takes an instance of `DataStoreConfiguration` as the user-provided settings in addition to the JSON values
* Added a field to the `Builder` called `ensureDefaults`. This field is only set to true if the new constructor is called.
* Modified `build()` method to populate defaults if  `ensureDefaults` is true

## Changes to `DataStoreConfiguration`
* Marked two builder methods(`builder(JSONObject)`, and `builder(JSONObject, DataStoreConfiguration)`) as package private so they should only be callable internally. This leaves the parameter-less `builder()` method as the only option when user's want to create an instance of `DataStoreConfiguration` to pass in overrides. When created via this method, the builder's `build()` method will leave any scalar fields not explicitly set by the user as null.
* Simplified getters and added a new getter for `getSyncIntervalInMinutes`